hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,234 @@
+
+func (factory *StorageTypeFactory) visitTypeName(_ *astmodel.TypeVisitor, name astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+
+	// Resolve the type name to the actual referenced type
+	actualDefinition, actualDefinitionFound := factory.types[name]
+
+	// Check for property specific handling
+	if vc.property != nil && actualDefinitionFound {
+
+		if et, ok := actualDefinition.Type().(*astmodel.EnumType); ok {
+			// Property type refers to an enum, so we use the base type instead
+			return et.BaseType(), nil
+		}
+	}
+
+	// Map the type name into our storage namespace
+	localRef, ok := name.PackageReference.AsLocalPackage()
+	if !ok {
+		return name, nil
+	}
+
+	storageRef := astmodel.MakeStoragePackageReference(localRef)
+	visitedName := astmodel.MakeTypeName(storageRef, name.Name())
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	_ *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+	oc := vc.forObject(object)
+
+	var errs []error
+	properties := object.Properties()
+	for i, prop := range properties {
+		p, err := factory.makeStorageProperty(prop, oc)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			properties[i] = p
+		}
+	}
+
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		return nil, err
+	}
+
+	ot := astmodel.NewObjectType().WithProperties(properties...)
+	return astmodel.NewStorageType(*ot), nil
+}
+
+// makeStorageProperty applies a conversion to make a variant of the property for use when
+// serializing to storage
+func (factory *StorageTypeFactory) makeStorageProperty(
+	prop *astmodel.PropertyDefinition,
+	pc StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error) {
+	for _, conv := range factory.propertyConversions {
+		p, err := conv(prop, pc.forProperty(prop))
+		if err != nil {
+			// Something went wrong, return the error
+			return nil, err
+		}
+		if p != nil {
+			// We have the conversion we need, return it promptly
+			return p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find a conversion for property %v", prop.PropertyName())
+}
+
+// Preserve properties required by the KubernetesResource interface unchanged as they're always required
+func (factory *StorageTypeFactory) preserveKubernetesResourceStorageProperties(
+	prop *astmodel.PropertyDefinition,
+	_ StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error) {
+	if prop.HasName(astmodel.AzureNameProperty) ||
+		prop.HasName(astmodel.OwnerProperty) {
+		// Keep these unchanged
+		return prop, nil
+	}
+
+	// No opinion, defer to another conversion
+	return nil, nil
+}
+
+func (factory *StorageTypeFactory) convertPropertiesForStorage(
+	prop *astmodel.PropertyDefinition,
+	ctx StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error) {
```

Comment:
This seems to not be used at all? Can it just be `_` then, or possibly removed entirely from the `propertyConversion` function signature since neither definition use it? (not sure removing it from func signature makes sense as maybe others might need it?)

Reply:
Oh, or is the parameter `prop` to the `Visit` call below supposed to be `ctx`?

Reply:
But if its supposed to be `ctx` and works as `prop` that seems to suggest it's not actually needed? I'm not actually sure how it works with prop as there are a bunch of casts in the visit handlers and I'd expect one of them to panic...

Reply:
Huh. This is one of those "I wonder how this ever worked" moments. In retesting, I find it does panic - should be passing `ctx` as the value to visit, not `prop`.